### ISSUE TITLE
Feature Request - Add IDs to Page Controls

### ISSUE THREAD

**Issue by CoHealer** [Azure/api-management-developer-portal#619]
Selenium and other UX test harnesses use unique control identifiers during test runs. Page controls in APIM do not have unique identifiers which makes it much more difficult to locate expected controls upon processing page responses. For example, the navigation link menu and it's child links do not have unique identifiers:

<ul class="nav">
--
  | 
  |  
  | 
  | <li class="nav-item">
  | <a href="/" class="nav-link nav-link-active">Home</a>
  | </li>
  | 
  |  
  | 
  |  
  | 
  | <li class="nav-item">
  | <a href="/apis" class="nav-link">APIs</a>
  | </li>
  | 
  |  
  | 
  |  
  | 
  | <li class="nav-item">
  | <a href="/products" class="nav-link">Products</a>
  | </li>
  | 
  | </ul>

While not completely blocking, tests have to be written using other less reliable ways of identifying these controls. This in turn could lead to churn in test code maintenance and perhaps even result in false test success/failure reporting.

**Comment by azaslonov** [comment]
Hi CoHealer, thanks for bringing this up.

Just to clarify, you're adding tests for the published website? Would auto-generated IDs (generated when component added to the page and never changed until component gets deleted) would be enough?

The reason I'm asking is because for a regular user this ID doesn't have any meaning if exposed in designer UI. Of course, there ways to solve it, but we'd like to consider options before implementing it, but autogeneration can be done fairly quickly.

**Comment by Ken-Stanfield** [comment]
Hi azaslonov - I am one of Cory's colleagues at VWAC. He asked me to weigh in on this issue.

Yes, an auto-generated ID (created when an element is added to a page which is then immutable) would work for our scenario. As I noted to Cory, a meaningful ID like `idHomePageLink` would be nice, but I suspect that you're talking about something that looks more like a partial GUID, like `rg94qz7`. This is certainly better than nothing.

If your proposal is for something like this, then I would restrict the length of the static, auto-generated ID to no more than 7 or 8 characters, maybe also restricted to the lowercase letters `a-z` and the numbers `0-9`.

I could then create a map of unfriendly IDs to meaningful human-readable IDs and I'd be good to go.

Having some kind of immutable ID (in whatever form) will certainly make things a lot easier for anyone writing Selenium integration or scenario-based tests for self-hosted instances of the APIM Developer Portal.

**Comment by azaslonov** [comment]
Alright, we'll see what we can do.

**Comment by Ken-Stanfield** [comment]
Thanks @azaslonov for at least considering this request. I know that it will make things A LOT easier for me in terms of writing Selenium-based automation tests for the APIM Developer Portal, and I suspect that I'm not alone.

I'm also hoping that whatever changes you make would apply both to the intrinsic UI pieces already in the Portal (e.g. those generated by Knockout or by other means such as generic menu bar items) as well as to custom pages and/or elements added by end-users.

**Comment by azaslonov** [comment]
Yes, every component bon every page/layout will have IDs. In fact, in Paperbits project we started working on end to end testing, and this goes not only to published portal, but to designers as well, so hopefully we will face/resolve most of possible issues even before you do.

**Comment by Ken-Stanfield** [comment]
Thanks for the update. That's great news!

**Comment by CoHealer** [comment]
@azaslonov, will this change be targeting your next planned release?

**Comment by azaslonov** [comment]
Can't promise next one specifically, but one of next ones. Now, we're switching back to weekly releases for the developer portal, so it should be much faster.

**Comment by CoHealer** [comment]
This is now three months old. Weekly releases is not improving the velocity. Is there an ETA for this or is the P2 label going to push this work down the stack?

**Comment by mikebudzynski** [comment]
@CoHealer Our priority is to address critical issues marked with the P0 and P1 labels, which are impacting all of our customers and affecting the most used functionalities of the developer portal. It's unlikely we will have capacity to address your request within the next month or even two. If you're working under strict timelines, I suggest that you consider contributing the changes to the codebase in this repository.

**Comment by AnRei123** [comment]
Not sure if the solution required in the here referenced ticket may also be relevant for a quick solution: [Deep links do not work after changing the permalink of the page or the layout - support a unique page ID suffix in the URL and a unique layout ID! #1106](https://github.com/Azure/api-management-developer-portal/issues/1106).
